File: upload-service/src/services/s3.ts
import { env } from "../config/env.js";
import { s3 } from "../config/s3.js";
import { getProfileKey } from "../utils/s3.js";
import { PutObjectCommand } from "@aws-sdk/client-s3";

export const uploadGoogleProfileImage = async ({
  id,
  url,
}: {
  id: number;
  url: string;
}) => {
  try {
    // Download image
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`Failed to download image: ${response.status}`);
    }

    const contentType = response.headers.get("content-type") || "image/jpeg";
    const arrayBuffer = await response.arrayBuffer();
    const buffer = Buffer.from(arrayBuffer);

    const key = getProfileKey({ id });

    // Upload to S3
    await s3.send(
      new PutObjectCommand({
        Bucket: env.AWS_S3_BUCKET,
        Key: key,
        Body: buffer,
        ContentType: contentType,
      }),
    );
  } catch (err) {
    console.log(err);
  }
};

File: user_service/src/user_service/services/s3.py
from user_service.config.settings import settings
from user_service.config.aws import s3_client


def create_presigned_upload_url(
    id: int,
    content_type: str,
) -> dict[str, str]:
    key = f"file/{id}"

    url = s3_client.generate_presigned_url(
        ClientMethod="put_object",
        Params={
            "Bucket": settings.aws_s3_bucket,
            "Key": key,
            "ContentType": content_type,
        },
        ExpiresIn=900,
    )

    return {
        "url": url,
        "key": key,
    }
